internal/vehicle/ford.go

```diff
@@ -134,36 +170,139 @@ func (v *Ford) vehicles() ([]string, error) {
 	return vehicles, err
 }
 
-// chargeState implements the api.Vehicle interface
-func (v *Ford) chargeState() (float64, error) {
-	var resp struct {
-		VehicleStatus struct {
-			Battery struct {
-				BatteryStatusActual struct {
-					Value int
-				}
+// Parses vehicle timestamp in format returned by Ford API and calculates age
+func (v *Ford) CalculateAge(timestamp string) (age time.Duration, err error) {
+	var timestampTime time.Time
+	const dateFormat = "01-02-2006 15:04:05"
+	timestampTime, err = time.Parse(dateFormat, timestamp)
+	age = time.Since(timestampTime)
+
+	return age, err
+}
+
+// VehicleStatus implements the /status response
+func (v *Ford) VehicleStatus() (res vehicleStatus, err error) {
+	uri := fmt.Sprintf("%s/api/vehicles/v3/%s/status", fordAPI, v.vin)
+
+	var req *http.Request
+	req, err = v.request(http.MethodGet, uri)
+	if err == nil {
+		err = v.DoJSON(req, &res)
+	}
+
+	var statusAge time.Duration
+	if err == nil {
+		statusAge, err = v.CalculateAge(res.VehicleStatus.LastRefresh)
+	}
+
+	if err == nil && statusAge > outdatedAfterMins*time.Minute {
+		// received data is considered outdated, server is requested to poll updated data from vehicle
+		v.log.DEBUG.Printf("Vehicle Status is outdated (age %v > %dm), requesting refresh", statusAge, outdatedAfterMins)
+		var updatedRes vehicleStatus
+		updatedRes, err = v.VehicleStatusRefresh()
+		if err == nil {
+			res = updatedRes
+			statusAge, err = v.CalculateAge(res.VehicleStatus.LastRefresh)
+			v.log.DEBUG.Printf("Refreshed Status Age: %v", statusAge)
+		}
+	}
+
+	return res, err
+}
+
+// Get updated vehicle status after requesting refresh
+func (v *Ford) VehicleStatusRefresh() (res vehicleStatus, err error) {
+	var commandId string
+	commandId, err = v.requestRefresh()
+
+	if err == nil {
+		uri := fmt.Sprintf("%s/api/vehicles/v3/%s/statusrefresh/%s", fordAPI, v.vin, commandId)
+
+		var req *http.Request
+
+		counter := 0
+		for counter < maxTrials {
+			req, err = v.request(http.MethodGet, uri)
+			if err == nil {
+				err = v.DoJSON(req, &res)
+			} else {
+				break
 			}
-			BatteryFillLevel struct {
-				Value float64
+
+			// if status is 200, the update has been completed
+			// otherwise request needs to be resent
+			if res.Status == 200 {
+				break
 			}
+
+			time.Sleep(1500 * time.Millisecond)
+			counter++
 		}
 
-		// "chargingStatus": { "value": "EvseNotDetected", ...
-		// "plugStatus": { "value": 1, ...
-		// "batteryChargeStatus": null, ...
+		if err == nil && counter >= maxTrials && res.Status != 200 {
+			err = fmt.Errorf("update of SoC not completed after timeout")
+			v.log.DEBUG.Print("update of SoC not completed after timeout")
```

Comment:
Either return the error or log it. Here: return "vehicle could not be updated"

Reply:
I decided to return the error, but handle it in the vehicleStatus function - a warning is logged and the cached status is returned